[SwiftUI] Support presenting JavaScript dialogs

rr-codes opened the pull request:
#### e2eb78275e5ee653431c9dedd94138f7c6efe03b
<pre>
[SwiftUI] Support presenting JavaScript dialogs
<a href="https://bugs.webkit.org/show_bug.cgi?id=284402">https://bugs.webkit.org/show_bug.cgi?id=284402</a>
<a href="https://rdar.apple.com/141108805">rdar://141108805</a>

Reviewed by Wenson Hsieh.

Add the ability to support JavaScript alerts, confirmation prompts, text input prompts, and file input.

* Source/WebKit/UIProcess/API/Swift/WKUIDelegateAdapter.swift: Added.
(WKUIDelegateAdapter.webView(_:runJavaScriptAlertPanelWithMessage:initiatedByFrame:)):
(WKUIDelegateAdapter.webView(_:runJavaScriptConfirmPanelWithMessage:initiatedByFrame:)):
(WKUIDelegateAdapter.webView(_:runJavaScriptTextInputPanelWithPrompt:defaultText:initiatedByFrame:)):
(WKUIDelegateAdapter.webView(_:runOpenPanelWith:initiatedByFrame:)):
* Source/WebKit/UIProcess/API/Swift/WebPage+DialogPresenting.swift: Added.
(DialogPresenting.handleJavaScriptAlert(_:initiatedBy:)):
(DialogPresenting.handleJavaScriptConfirm(_:initiatedBy:)):
(DialogPresenting.handleJavaScriptPrompt(_:defaultText:initiatedBy:)):
(DialogPresenting.handleFileInputPrompt(_:initiatedBy:)):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPage_v0.backingWebView):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/SwiftBrowser/Source/BrowserViewModel.swift:
(hash(into:)):
(owner):
(handleJavaScriptAlert(_:initiatedBy:)):
(handleJavaScriptConfirm(_:initiatedBy:)):
(handleJavaScriptPrompt(_:defaultText:initiatedBy:)):
(handleFileInputPrompt(_:initiatedBy:)):
(BrowserViewModel.currentDialog):
(BrowserViewModel.currentFilePicker):
(BrowserViewModel.didImportFiles(_:any:)):
* Tools/SwiftBrowser/Source/ContentView.swift:
(DialogActionsView.body):
(DialogMessageView.body):
(ContentView.body):

Canonical link: <a href="https://commits.webkit.org/287773@main">https://commits.webkit.org/287773@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/12ad3f84b251c72d9ec696f605985d02ee5c159e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80778 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/300 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34714 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/85305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/31762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82889 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/317 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8098 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/85305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/31762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83847 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/146 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/73530 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/85305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/72 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/27690 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/30219 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/71638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/28222 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/86737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/8005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/86737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/8182 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/69366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/86737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/13588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12528 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7967 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/13488 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/7806 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/11325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9612 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lilyspiniolas commented on Tools/SwiftBrowser/Source/ContentView.swift:
```diff
@@ -55,6 +55,67 @@ fileprivate struct ToolbarBackForwardMenuView: View {
+        case let .confirm(_, dismissConfirm):
+            Button("OK") {
+                dismissConfirm(.ok)
+            }
```
Should `.keyboardShortcut(.defaultAction)` be applied here too?

rr-codes replied:
it's not needed, since if there are only buttons involved, SwiftUI can figure out which one is the default if the other is cancel.

axiixc replied:
Yeah, if you had more than one non-cancel non-destructive role button you'd probably need that to tie break otherwise.

lilyspiniolas commented on Tools/SwiftBrowser/Source/ContentView.swift:
```diff
@@ -55,6 +55,67 @@ fileprivate struct ToolbarBackForwardMenuView: View {
+        case let .alert(_, dismissAlert):
+            Button("Dismiss") {
+                dismissAlert()
+            }
```
nit: The current button for JS alerts is "Close," parity might be good here? Also not sure if it matters

rr-codes replied:
ya it doesn't really matter since this is just a test app, but I can change it

axiixc commented on Source/WebKit/UIProcess/API/Swift/WKUIDelegateAdapter.swift:
```diff
@@ -0,0 +1,77 @@
+import Foundation
+internal import WebKit_Private
+
+fileprivate struct DefaultDialogPresenting: DialogPresenting {
```
Nit `fileprivate` is just `private` at top level scope usually.

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage+DialogPresenting.swift:
```diff
@@ -0,0 +1,88 @@
+
+    @MainActor
+    func handleJavaScriptConfirm(message: String, initiatedBy frame: WebPage_v0.FrameInfo) async -> WebPage_v0.JavaScriptConfirmResult {
+        .cancel
```
I'd probably prefer to write out `return` for a full method like this. But that's really just a stylistic thing to decide on.

rr-codes replied:
I think I use implicit returns in other cases, so I think I'll leave it as-is for now

axiixc commented on Source/WebKit/UIProcess/API/Swift/WKUIDelegateAdapter.swift:
```diff
@@ -0,0 +1,77 @@
+    func webView(_ webView: WKWebView, runJavaScriptConfirmPanelWithMessage message: String, initiatedByFrame frame: WKFrameInfo) async -> Bool {
+        let result = await dialogPresenter.handleJavaScriptConfirm(message: message, initiatedBy: .init(frame))
+        return switch result {
+        case .ok:
```
Since you are using implicit returns already, you could also shorten this to one line. That's kinda been my logic for these:
- Multiple lines, more complex, write `return`
- Single line, simpler expression, elide it

¯\\\_(ツ)\_/¯

rr-codes replied:
so it would be like

```
switch await dialogPresenter... { 
```

?

axiixc replied:
Sorry, I was implying:
```swift
let result = await dialogPresenter.handleJavaScriptConfirm(message: message, initiatedBy: .init(frame))

return switch result {
case .ok: true
case .cancel: false
}
```

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage+DialogPresenting.swift:
```diff
@@ -0,0 +1,88 @@
+// MARK: Supporting types
+
+extension WebPage_v0 {
+    public enum JavaScriptConfirmResult: Sendable {
```
IIRC `enum`s are also implicitly `Hashable`, even as `public`, so you might want to write that down to be explicit. This is a bit of an outlier for Swift's general philosophy on API, and can bite you if were to add a new non-Hashable case like `case foo(OopsType)`. If you write down `: Hashable` now then you'll at least turn that into a build time error in the future (though we should also be running the ABI checker which would catch this _eventually_ too).

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage+DialogPresenting.swift:
```diff
@@ -0,0 +1,88 @@
+
+// MARK: Default implementation
+
+@_spi(Private)
```
I think this is technically unnecessary (the type itself is already SPI so shouldn't all extensions on it?)

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -34,23 +34,33 @@ public class WebPage_v0 {
 
-    private init(configuration: Configuration, _navigationDecider navigationDecider: (any NavigationDeciding)?) {
-        self.configuration = configuration
+    private init(_configuration: Configuration, _navigationDecider navigationDecider: (any NavigationDeciding)?, _dialogPresenter dialogPresenter: (any DialogPresenting)?) {
```
Curious about the `_`s in an already `private` method? Is this to avoid ambiguous overloads?

rr-codes replied:
ya, the initializers beneath it caused this to become ambiguous

axiixc commented on Tools/SwiftBrowser/Source/BrowserViewModel.swift:
```diff
@@ -39,17 +39,87 @@ extension PDF: Transferable {
+
+    func handleJavaScriptAlert(message: String, initiatedBy frame: WebPage_v0.FrameInfo) async {
+        await withCheckedContinuation { continuation in
+            owner?.currentDialog = Dialog(configuration: .alert(message) {
```
Super nit, but you should be able to just pass this as an unapplied function rather than making a closure just to call it like a thunk:
```suggestion
            owner?.currentDialog = Dialog(configuration: .alert(message, continuation.resume))
```

axiixc commented on Tools/SwiftBrowser/Source/BrowserViewModel.swift:
```diff
@@ -39,17 +39,87 @@ extension PDF: Transferable {
+
+    func handleJavaScriptConfirm(message: String, initiatedBy frame: WebPage_v0.FrameInfo) async -> WebPage_v0.JavaScriptConfirmResult {
+        await withCheckedContinuation { continuation in
+            owner?.currentDialog = Dialog(configuration: .confirm(message) {
```
And likewise
```suggestion
            owner?.currentDialog = Dialog(configuration: .confirm(message, continuation.resume(returning:)))
```

rr-codes replied:
When I try this, I get

```
Converting a value of type '@Sendable (sending WebPage_v0.JavaScriptConfirmResult) -> ()' to type '@Sendable (WebPage_v0.JavaScriptConfirmResult) -> Void' risks causing data races
```

rr-codes replied:
I made the enum associated value have a `sending` parameter and that fixed it

axiixc replied:
Hmm, I was expecting this to just convert, but some of the concurrency annotations are more strict (ex: you can generally pass a `(Int?) -> Void` to an `(Int) -> Void` because it's a subtype).

axiixc commented on Tools/SwiftBrowser/Source/BrowserViewModel.swift:
```diff
@@ -67,6 +137,38 @@ final class BrowserViewModel {
+        }
+    }
+
+    @ObservationIgnored
```
Why ignored?

rr-codes replied:
I figured that it wasn't necessary since the "isPresentingDialog" property takes care of the observable-ness needed

axiixc replied:
I'd personally avoid overeager optimizations. If you capture two dependencies and they both change in one render cycle that's still one re-render. Conversely if you only capture one and the other changes you'll have a correctness bug.

`@ObservationIgnored` is good when you want to semantically have no side effects on a read because it's transparent to the caller for whatever reason. Usually I use this with memoized subscripts or methods, because the fact that you've cached a new value doesn't mean you want a dependency any time you cache any other new value, unless of course you have a debug view showing the contents of the cache in which case it _would_ be wrong.

axiixc commented on Tools/SwiftBrowser/Source/BrowserViewModel.swift:
```diff
@@ -106,4 +208,18 @@ final class BrowserViewModel {
+        case let .success(urls):
+            currentFilePicker!.completion(.ok(urls))
+
+        case .failure(_):
```
You can also drop the `()` entirely then:
```suggestion
        case .failure:
```

axiixc commented on Tools/SwiftBrowser/Source/ContentView.swift:
```diff
@@ -55,6 +55,67 @@ fileprivate struct ToolbarBackForwardMenuView: View {
+    var body: some View {
+        switch dialog.configuration {
+        case let .alert(_, dismissAlert):
+            Button("Dismiss") {
```
```suggestion
            Button("Dismiss", action: dismissAlert)
```

axiixc commented on Tools/SwiftBrowser/Source/ContentView.swift:
```diff
@@ -55,6 +55,67 @@ fileprivate struct ToolbarBackForwardMenuView: View {
+
+    var body: some View {
+        switch dialog.configuration {
+        case let .alert(message, _):
```
This is just sample code, so N.B.D., but another way to handle this would be making a `var message: String { … }` extension on the enum so that you could write `Text(dialog.configuration.message)` here. That's mostly isomorphic, but it does simplify your view tree. That only applies because the types here are all otherwise identical, so if there were any other customizations I probably wouldn't recommend it. Just throwing it out.

axiixc commented on Tools/SwiftBrowser/Source/BrowserViewModel.swift:
```diff
@@ -39,17 +39,87 @@ extension PDF: Transferable {
+        let configuration: Configuration
+
+        func hash(into hasher: inout Hasher) {
+            self.id.hash(into: &hasher)
```
Almost missed this, but what you actually want is
```suggestion
           hasher.combine(id)
```

rr-codes committed:
[SwiftUI] Support presenting JavaScript dialogs
https://bugs.webkit.org/show_bug.cgi?id=284402
rdar://141108805

Reviewed by Wenson Hsieh.

Add the ability to support JavaScript alerts, confirmation prompts, text input prompts, and file input.

* Source/WebKit/UIProcess/API/Swift/WKUIDelegateAdapter.swift: Added.
(WKUIDelegateAdapter.webView(_:runJavaScriptAlertPanelWithMessage:initiatedByFrame:)):
(WKUIDelegateAdapter.webView(_:runJavaScriptConfirmPanelWithMessage:initiatedByFrame:)):
(WKUIDelegateAdapter.webView(_:runJavaScriptTextInputPanelWithPrompt:defaultText:initiatedByFrame:)):
(WKUIDelegateAdapter.webView(_:runOpenPanelWith:initiatedByFrame:)):
* Source/WebKit/UIProcess/API/Swift/WebPage+DialogPresenting.swift: Added.
(DialogPresenting.handleJavaScriptAlert(_:initiatedBy:)):
(DialogPresenting.handleJavaScriptConfirm(_:initiatedBy:)):
(DialogPresenting.handleJavaScriptPrompt(_:defaultText:initiatedBy:)):
(DialogPresenting.handleFileInputPrompt(_:initiatedBy:)):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPage_v0.backingWebView):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/SwiftBrowser/Source/BrowserViewModel.swift:
(hash(into:)):
(owner):
(handleJavaScriptAlert(_:initiatedBy:)):
(handleJavaScriptConfirm(_:initiatedBy:)):
(handleJavaScriptPrompt(_:defaultText:initiatedBy:)):
(handleFileInputPrompt(_:initiatedBy:)):
(BrowserViewModel.currentDialog):
(BrowserViewModel.currentFilePicker):
(BrowserViewModel.didImportFiles(_:any:)):
* Tools/SwiftBrowser/Source/ContentView.swift:
(DialogActionsView.body):
(DialogMessageView.body):
(ContentView.body):

Canonical link: https://commits.webkit.org/287773@main
